### ISSUE TITLE
Cannot create new boards in latest revision on Sandstorm!

### ISSUE THREAD

**Issue by xarinatan** [wekan/wekan#1398]
## Issue
**Impacted version**: 0.60.0~2017-11-29

**Server Setup Information**: 

* Operating System: Arch with latest updates
* Deployment Method(tar/sandstorm/snap/docker/etc /http frontend (nginx, httpd, ...)): Sandstorm
* Node Version: Sandstorm builtin (host has no nodeJS)
* MongoDB Version: Sandstorm builtin
* ROOT_URL environment variable (Is there a subfolder?):

**Problem description**:
When creating a new board in my sandstorm server, it fails to load it, stating "Error: remote exception: Peer disconnected.
C++ location:(remote):??
type: disconnected"
The console log reveals it's missing a module, but I'm not sure that's the issue: 
```
child process started successfully, parent exiting
** Starting Meteor...
{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version

/programs/server/node_modules/fibers/future.js:313
						throw(ex);
						^
MongoError: failed to connect to server [127.0.0.1:4001] on first connect
    at Object.Future.wait (/programs/server/node_modules/fibers/future.js:449:15)
    at new MongoConnection (packages/mongo/mongo_driver.js:211:27)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Mongo.Collection (packages/mongo/collection.js:99:40)
    at packages/idmontie_migrations.js:38:15
    at packages/idmontie_migrations.js:255:4
    at packages/idmontie_migrations.js:267:3
    at /programs/server/boot.js:303:34
    - - - - -
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:326:35)
    at emitOne (events.js:77:13)
    at [object Object].emit (events.js:169:7)
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:270:12)
    at [object Object].g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at [object Object].emit (events.js:172:7)
    at Socket.<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:175:49)
    at Socket.g (events.js:260:16)
    at emitOne (events.js:77:13)
** HTTP-BRIDGE: App server exited with status code: 1
```

I'm using the latest version Sandstorm provides and I've seen this behavior across multiple accounts. 
Existing boards are unaffected, but it'd be really nice to be able to create new ones ;)

**Comment by ocdtrekkie** [comment]
Intriguing. I can definitely create new boards with Wekan 0.60 on Sandstorm Oasis.

**Comment by xarinatan** [comment]
That's intriguing indeed, I guess it must be on my side then.. not sure what's up, I have full ECC hardware and I'm running BTRFS as filesystem, I also tried to restart sandstorm, nothing really seemed to fix it so I thought it must be in the application itself

**Comment by xarinatan** [comment]
I heard Oasis runs a slightly different version of sandstorm, could that be related perhaps?

**Comment by ocdtrekkie** [comment]
Not usually, but I just double checked that it worked on Sandstorm Alpha for new boards as well.

Hmm... the usual is try a different browser and such. I have a new Wekan build to review and release, which might fix it, if it's something wrong with the package on your system.

**Comment by ocdtrekkie** [comment]
If you want, I've rudimentarily checked that 0.62 both works on new grains and existing grains: https://apps.sandstorm.io/app/m86q05rdvj14yvn78ghaxynqz7u2svw6rnttptxx49g1785cdv1h?experimental=true

**Comment by xarinatan** [comment]
Thanks for the experimental build and tips, but I'm afraid they didn't solve the issue :( I have the same issue with Chrome on Windows as well as Firefox on Linux which I originally tried it with. A friend is observing the same issue on his account/browser (I think chrome on windows as well). 
The website hangs for ~15 seconds before failing, and since the logs say it failed to connect to mongodb, I assume that fails to start? Do you know of a way to get the mongodb logs from sandstorm perhaps? I have full root access to the machine

**Comment by ocdtrekkie** [comment]
Each grain has it's own mongodb. If you download a grain backup, the log for it might be in there somewhere, if it saves it. It's hard to even picture what could be causing your issue.

cc: @kentonv I am baffled, if you happen to notice this tag, and have any ideas.

**Comment by xarinatan** [comment]
Right, of course, thanks! I attached an empty wekan board's backup which wouldn't load on my server. 
Looking at the logs it seems that it just gets stuck booting up? the I/O isn't terribly slow or anything, rebooting the entire VM took less than 30 seconds, so no idea what could cause this. It was working fine before.. 

[Untitled Wekan board.zip](https://github.com/wekan/wekan/files/1565339/Untitled.Wekan.board.zip)

**Comment by xet7** [comment]
@xarinatan 

Your board backup did work normally for me with Wekan v0.62 on local Sandstorm dev. No problems. Logs:

```
...** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: userns
** Starting Mongo...
2017-12-17T14:19:33.085+0000 I STORAGE  Engine custom option: log=(prealloc=false,file_max=200KB)
about to fork child process, waiting until server is ready for connections.
forked process: 10
child process started successfully, parent exiting
** Starting Meteor...
> Starting board-background-color migration.
> Finishing board-background-color migration.
> Starting lowercase-board-permission migration.
> Finishing lowercase-board-permission migration.
> Starting change-attachments-type-for-non-images migration.
> Finishing change-attachments-type-for-non-images migration.
> Starting card-covers migration.
> Finishing card-covers migration.
> Starting use-css-class-for-boards-colors migration.
> Finishing use-css-class-for-boards-colors migration.
> Starting denormalize-star-number-per-board migration.
> Finishing denormalize-star-number-per-board migration.
> Starting add-member-isactive-field migration.
> Finishing add-member-isactive-field migration.
> Starting add-sort-checklists migration.
> Finishing add-sort-checklists migration.
> Starting enforce-public-visibility-for-sandstorm migration.
> Finishing enforce-public-visibility-for-sandstorm migration.
```

You could try removing Wekan App and adding newest Wekan App again:
https://apps.sandstorm.io/app/m86q05rdvj14yvn78ghaxynqz7u2svw6rnttptxx49g1785cdv1h?experimental=true

I'm unable to duplicate this issue.

**Comment by xet7** [comment]
This message means that there is no binary version of bson extension, so it is using javascript version instead. It does not prevent Wekan from working, it can be ignored.

```
js-bson: Failed to load c++ bson extension, using pure JS version
```

I have this fixed in newer Meteor 1.6 version of Wekan that does not work on Sandstorm yet. I always test that Wekan does start normally on Sandstorm when I'm publishing new Sandstorm version.

**Comment by xet7** [comment]
Inside your Untitled Wekan board backup .zip file are these logs, I don't know why there is messages about unable to connect to MongoDB:

```
** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: privileged
** Starting Mongo...
2017-12-17T04:47:11.190+0000 I STORAGE  Engine custom option: log=(prealloc=false,file_max=200KB)
about to fork child process, waiting until server is ready for connections.
forked process: 10
child process started successfully, parent exiting
** Starting Meteor...

/programs/server/node_modules/fibers/future.js:313
						throw(ex);
						^
MongoError: failed to connect to server [127.0.0.1:4001] on first connect
    at Object.Future.wait (/programs/server/node_modules/fibers/future.js:449:15)
    at new MongoConnection (packages/mongo/mongo_driver.js:211:27)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Mongo.Collection (packages/mongo/collection.js:99:40)
    at packages/idmontie_migrations.js:38:15
    at packages/idmontie_migrations.js:255:4
    at packages/idmontie_migrations.js:267:3
    at /programs/server/boot.js:303:34
    - - - - -
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:326:35)
    at emitOne (events.js:77:13)
    at [object Object].emit (events.js:169:7)
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:270:12)
    at [object Object].g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at [object Object].emit (events.js:172:7)
    at Socket.<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:175:49)
    at Socket.g (events.js:260:16)
    at emitOne (events.js:77:13)
** HTTP-BRIDGE: App server exited with status code: 1
** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: privileged
** Starting Mongo...
2017-12-17T04:47:17.524+0000 I STORAGE  Engine custom option: log=(prealloc=false,file_max=200KB)
about to fork child process, waiting until server is ready for connections.
forked process: 10
child process started successfully, parent exiting
** Starting Meteor...

/programs/server/node_modules/fibers/future.js:313
						throw(ex);
						^
MongoError: failed to connect to server [127.0.0.1:4001] on first connect
    at Object.Future.wait (/programs/server/node_modules/fibers/future.js:449:15)
    at new MongoConnection (packages/mongo/mongo_driver.js:211:27)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Mongo.Collection (packages/mongo/collection.js:99:40)
    at packages/idmontie_migrations.js:38:15
    at packages/idmontie_migrations.js:255:4
    at packages/idmontie_migrations.js:267:3
    at /programs/server/boot.js:303:34
    - - - - -
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:326:35)
    at emitOne (events.js:77:13)
    at [object Object].emit (events.js:169:7)
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:270:12)
    at [object Object].g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at [object Object].emit (events.js:172:7)
    at Socket.<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:175:49)
    at Socket.g (events.js:260:16)
    at emitOne (events.js:77:13)
** HTTP-BRIDGE: App server exited with status code: 1
** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: privileged
** Starting Mongo...
2017-12-17T04:48:59.585+0000 I STORAGE  Engine custom option: log=(prealloc=false,file_max=200KB)
about to fork child process, waiting until server is ready for connections.
forked process: 10
child process started successfully, parent exiting
** Starting Meteor...

/programs/server/node_modules/fibers/future.js:313
						throw(ex);
						^
MongoError: failed to connect to server [127.0.0.1:4001] on first connect
    at Object.Future.wait (/programs/server/node_modules/fibers/future.js:449:15)
    at new MongoConnection (packages/mongo/mongo_driver.js:211:27)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Mongo.Collection (packages/mongo/collection.js:99:40)
    at packages/idmontie_migrations.js:38:15
    at packages/idmontie_migrations.js:255:4
    at packages/idmontie_migrations.js:267:3
    at /programs/server/boot.js:303:34
    - - - - -
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:326:35)
    at emitOne (events.js:77:13)
    at [object Object].emit (events.js:169:7)
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:270:12)
    at [object Object].g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at [object Object].emit (events.js:172:7)
    at Socket.<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:175:49)
    at Socket.g (events.js:260:16)
    at emitOne (events.js:77:13)
** HTTP-BRIDGE: App server exited with status code: 1
** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: privileged
** Starting Mongo...
2017-12-17T04:49:05.466+0000 I STORAGE  Engine custom option: log=(prealloc=false,file_max=200KB)
about to fork child process, waiting until server is ready for connections.
forked process: 10
child process started successfully, parent exiting
** Starting Meteor...

/programs/server/node_modules/fibers/future.js:313
						throw(ex);
						^
MongoError: failed to connect to server [127.0.0.1:4001] on first connect
    at Object.Future.wait (/programs/server/node_modules/fibers/future.js:449:15)
    at new MongoConnection (packages/mongo/mongo_driver.js:211:27)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Mongo.Collection (packages/mongo/collection.js:99:40)
    at packages/idmontie_migrations.js:38:15
    at packages/idmontie_migrations.js:255:4
    at packages/idmontie_migrations.js:267:3
    at /programs/server/boot.js:303:34
    - - - - -
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:326:35)
    at emitOne (events.js:77:13)
    at [object Object].emit (events.js:169:7)
    at [object Object].<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:270:12)
    at [object Object].g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at [object Object].emit (events.js:172:7)
    at Socket.<anonymous> (/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:175:49)
    at Socket.g (events.js:260:16)
    at emitOne (events.js:77:13)
** HTTP-BRIDGE: App server exited with status code: 1
```

**Comment by kentonv** [comment]
The backup contains mongo.log but unfortunately it's not very informative:

    2017-12-17T04:49:05.502+0000 W -        [initandlisten] Detected unclean shutdown - /var/wiredTigerDb/mongod.lock is not empty.
    2017-12-17T04:49:05.502+0000 W STORAGE  [initandlisten] Recovering data from the last clean checkpoint.
    2017-12-17T04:49:05.502+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=1G,session_max=20000,eviction=(threads_max=4),statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),log=(prealloc=false,file_max=200KB)
    2017-12-17T04:49:07.337+0000 I STORAGE  [initandlisten] dropping unused ident: collection-0-4768632667234777734
    2017-12-17T04:49:08.609+0000 I CONTROL  [initandlisten] MongoDB starting : pid=10 port=4001 dbpath=/var/wiredTigerDb 64-bit host=sandbox
    2017-12-17T04:49:08.610+0000 I CONTROL  [initandlisten] db version v3.0.7-pre-
    2017-12-17T04:49:08.610+0000 I CONTROL  [initandlisten] git version: nogitversion
    2017-12-17T04:49:08.610+0000 I CONTROL  [initandlisten] build info: Linux zero 4.1.0-1-amd64 #1 SMP Debian 4.1.3-1 (2015-08-03) x86_64 BOOST_LIB_VERSION=1_49
    2017-12-17T04:49:08.610+0000 I CONTROL  [initandlisten] allocator: tcmalloc
    2017-12-17T04:49:08.610+0000 I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1", http: { enabled: false }, port: 4001 }, processManagement: { fork: true }, security: { authorization: "disabled" }, storage: { dbPath: "/var/wiredTigerDb", engine: "wiredTiger", wiredTiger: { engineConfig: { cacheSizeGB: 1, configString: "log=(prealloc=false,file_max=200KB)" } } }, systemLog: { destination: "file", path: "/var/wiredTigerDb/mongo.log" } }

... it just stops there.

My guess is that Mongo is crashing for some reason that's specific to the system. It could be related to the specific kernel version, or the specific CPU (but more likely the kernel).

What kernel version are you running? Try: `uname -a`

**Comment by xarinatan** [comment]
Sorry for the late reply, I've been really busy the past few days. 
Good call @kentonv ! I'm running 4.14.6-1-hardened #1 SMP PREEMPT Fri Dec 15 02:35:54 CET 2017 x86_64 GNU/Linux on Arch, I kind of forgot about running that hardened kernel, but I never expected it to get in the way like this (it's a GRsec derivative kernel afaik).
Running without the hardened kernel solves my issue!

![screenshot_20171218_184742](https://user-images.githubusercontent.com/1271027/34119890-06b47182-e424-11e7-82f4-0456333d8418.png)
Thanks! And I'm sorry for the scary title, I really thought I tried everything :P 
Do you have any idea what part of the hardened kernel could be causing this? I'll contact them to let them know about this issue

**Comment by kentonv** [comment]
@xarinatan Glad to hear you figured it out.

It's hard to say what about the hardened kernel could be an issue, since Mongo didn't report any particular error. We'd probably have to run it in a debugger or strace to find out what exactly it did leading up to the failure. It is surprising that Sandstorm itself (which uses Mongo, after all) worked fine, and only inside Wekan did something go wrong.

**Comment by xarinatan** [comment]
just a note to this old thread; I just found that I can more or less reproduce this issue by pausing/saving the VM that Sandstorm is running on and keeping it paused for a while. For some reason most things work well (using existing boards, creating new grains of a lot of other apps, etc) but creating new Wekan boards causes the original error. Rebooting doesn't fix it, but turning off the VM and then starting it up again does.. The only thing I can think of is time/clock issues, but I'm running ntpd on the VM so I'm not sure what's up..  It's a very peculiar thing :thinking: it's not really an issue anymore though, just thought I'd share my findings :)

**Comment by ocdtrekkie** [comment]
@xarinatan I can at least say this may be the most intriguing bug report yet, so by all means, keep us up to date.

**Comment by xarinatan** [comment]
@ocdtrekkie :) I will if I figure out more, it's a very strange thing that's been haunting me for a long time, especially since I use Wekan a lot. 
My setup is pretty straight forward; Arch Linux, KVM/qemu + Libvirtd + virt-manager and kimchi. I used to run a hardened kernel but I haven't done so since I ran into this issue. 
Recently the issue showed up again after I did maintenance and I couldn't figure out why, so after several reboots I turned the VM off completely and checked if the CPU settings were correct, then without changing anything started it up again and somehow that had fixed the issue. I know that KVM preserves the state of a VM's hardware when a VM's guest OS reboots, so one of the virtual devices, presumably the RTC or some other timer, must've gone haywire when the VM was paused and the host was rebooted.
